trackpy/packing.py

```diff
@@ -0,0 +1,130 @@
+    """
+    Calculate the pair correlation function in 2 dimensions.
+
+    Parameters
+    ----------
+    feat : Pandas DataFrame
+        DataFrame containing the x and y coordinates of particles
+    fration : float, optional
+        The fraction of particles to calculate g(r) with. May be used to increase speed of function. Particles selected at random.
+    dr : float, optional
+        The bin width
+    cutoff : float, optional
+        Maximum distance to calculate g(r)
+    ndensity : float, optional
+        Density of particle packing. If not specified, density will be calculated assuming rectangular homogenous arangement
+    boundary : Tuple, optional
+        Tuple specifying rectangular boundary of partcicles (xmin, xmax, ymin, ymax)
+
+    Returns
+    -------
+    r_edges : array
+        Return the bin edges
+    g_r : array
+        The values of g_r
+    """
+
+    if boundary is None:
+        warn("Rectangular packing is assumed. Boundaries are determined by edge particles.")
+        xmin = feat.x.min()
+        xmax = feat.x.max()
+        ymin = feat.y.min()
+        ymax = feat.y.max()
+    else:
+        xmin = boundary[0]
+        xmax = boundary[1]
+        ymin = boundary[2]
+        ymax = boundary[3]
+
+    if ndensity is None:
+        ndensity = feat.x.count() / ((feat.x.max() - feat.x.min()) * (feat.y.max() - feat.y.min())) #particle packing density
+        warn("Rectangular homogenous packing is assumed to calculate particle density.")
+
+
+    p_indexes = random.sample(range(len(feat)), int(fraction*len(feat))) #grab random sample of particles
+    r_edges = np.arange(dr, cutoff + 2*dr, dr) #radii to search for particles
+    g_r = np.zeros(len(r_edges) - 1)
+    max_p_count =  int(np.pi * (r_edges.max() + dr)**2 * ndensity * 10) #upper bound for neighborhood particle count
+    ckdtree = cKDTree(feat[['x', 'y']])#initialize kdtree
+    points = feat.as_matrix(['x', 'y'])
+
+    #create reference disks of all radii in r_edges
+    n = 50
+    x = np.linspace(-1, 1, n)
+    y = x.copy()
+    x = np.tile(x, n)
+    y = y.repeat(n)
+    mask1 = x**2 + y**2 <= 1
+    mask2 = x**2 + y**2 >= 1 - dr
+    x = x[mask1 & mask2]
+    y = y[mask1 & mask2]
+    refx = np.ones((len(r_edges), len(x)))
+    refx[:] = x
+    refx *= r_edges.repeat(len(x)).reshape((len(r_edges), len(x)))
+    refy = np.ones((len(r_edges), len(y)))
+    refy[:] = y
+    refy *= r_edges.repeat(len(y)).reshape((len(r_edges), len(y)))
+
+    for idx in p_indexes:
+        dist, idxs = ckdtree.query(points[idx], k=max_p_count, distance_upper_bound=(cutoff+dr))
+        area = np.ones(len(r_edges))
+
+        ##Find the number of edge collisions at each radii
+        collisions = num_wall_collisions(points[idx], r_edges, xmin, xmax, ymin, ymax)
+
+        #if some disk will collide with the wall, we need to implement edge handling
+        if collisions.max() > 0:
```

Comment:
Isn't `collisions` a scalar? (Also: could you do this check for all points at once?)